File: reference/python/phisql/catalog.py
# Copyright 2026 Philterd, LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
In-memory view of the PhiSQL spec catalog YAML files.

The catalog files under ``spec/v1.0/catalog/`` are the single source of truth
for entity types, strategies, keywords, and predicate forms. The compiler is
driven entirely by them; this module loads the subset the compiler needs
(entity types and strategies). Lookups by name are case-insensitive, matching
the language spec.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

import yaml

from . import _paths


@dataclass(frozen=True)
class StrategyArg:
    """A named argument allowed on a strategy."""

    name: str
    phileas_field: Optional[str]
    type: Optional[str]
    enum_values: tuple = ()


@dataclass(frozen=True)
class Strategy:
    """Catalog entry for a filter strategy."""

    name: str
    phileas_enum: str
    args: tuple = ()

    def find_arg(self, arg_name: Optional[str]) -> Optional[StrategyArg]:
        """Returns the strategy argument with the given name, or None."""
        if arg_name is None:
            return None
        for arg in self.args:
            if arg.name.lower() == arg_name.lower():
                return arg
        return None


@dataclass(frozen=True)
class EntityType:
    """Catalog entry for an entity type."""

    name: str
    phileas_field: str
    phileas_strategies_field: str


class Catalog:
    """Loaded entity-type and strategy catalogs, keyed case-insensitively."""

    #: Catalog version (matches the spec version this catalog targets).
    VERSION = "v1.0"

    def __init__(self, entities_by_name: dict, strategies_by_name: dict):
        self._entities_by_name = entities_by_name
        self._strategies_by_name = strategies_by_name

    @classmethod
    def load_default(cls) -> "Catalog":
        """Loads the v1.0 catalog from the spec ``catalog/`` directory."""
        catalog_dir = _paths.catalog_dir()

        entities = {}
        with (catalog_dir / "entity-types.yaml").open(encoding="utf-8") as fh:
            root = yaml.safe_load(fh)
        for item in root.get("entities") or []:
            entity = EntityType(
                name=item["name"],
                phileas_field=item["phileas_field"],
                phileas_strategies_field=item["phileas_strategies_field"],
            )
            entities[entity.name.upper()] = entity

        strategies = {}
        with (catalog_dir / "strategies.yaml").open(encoding="utf-8") as fh:
            root = yaml.safe_load(fh)
        for item in root.get("strategies") or []:
            args = []
            for raw in item.get("args") or []:
                enum_values = raw.get("enum_values") or []
                args.append(
                    StrategyArg(
                        name=raw["name"],
                        phileas_field=raw.get("phileas_field"),
                        type=raw.get("type"),
                        enum_values=tuple(enum_values),
                    )
                )
            strategy = Strategy(
                name=item["name"],
                phileas_enum=item["phileas_enum"],
                args=tuple(args),
            )
            strategies[strategy.name.upper()] = strategy

        return cls(entities, strategies)

    def get_entity(self, name: Optional[str]) -> Optional[EntityType]:
        """Returns the entity type with the given (case-insensitive) name, or None."""
        if name is None:
            return None
        return self._entities_by_name.get(name.upper())

    def get_strategy(self, name: Optional[str]) -> Optional[Strategy]:
        """Returns the strategy with the given (case-insensitive) name, or None."""
        if name is None:
            return None
        return self._strategies_by_name.get(name.upper())

File: reference/dotnet/PhiSql/Catalog.cs
// Copyright 2026 Philterd, LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;

namespace Philterd.PhiSql;

/// <summary>A named argument allowed on a strategy.</summary>
public sealed record StrategyArg(string Name, string? PhileasField, string? Type, IReadOnlyList<string> EnumValues);

/// <summary>Catalog entry for a filter strategy. <c>DateOnly</c> marks date-only strategies.</summary>
public sealed record Strategy(string Name, string PhileasEnum, IReadOnlyList<StrategyArg> Args, bool DateOnly = false)
{
    /// <summary>Returns the strategy argument with the given (case-insensitive) name, or null.</summary>
    public StrategyArg? FindArg(string? argName)
    {
        if (argName is null) return null;
        foreach (StrategyArg arg in Args)
            if (string.Equals(arg.Name, argName, StringComparison.OrdinalIgnoreCase))
                return arg;
        return null;
    }
}

/// <summary>Catalog entry for an entity type.</summary>
public sealed record EntityType(string Name, string PhileasField, string PhileasStrategiesField);

/// <summary>
/// In-memory view of the PhiSQL spec catalog YAML files. The catalog is the
/// single source of truth for entity types and strategies; the compiler is
/// driven entirely by it. Lookups by name are case-insensitive.
/// </summary>
public sealed class Catalog
{
    /// <summary>Catalog version (matches the spec version this catalog targets).</summary>
    public const string Version = "v1.0";

    private readonly IReadOnlyDictionary<string, EntityType> _entitiesByName;
    private readonly IReadOnlyDictionary<string, Strategy> _strategiesByName;

    private Catalog(IReadOnlyDictionary<string, EntityType> entities,
                    IReadOnlyDictionary<string, Strategy> strategies)
    {
        _entitiesByName = entities;
        _strategiesByName = strategies;
    }

    /// <summary>Loads the v1.0 catalog from the YAML embedded in this assembly.</summary>
    public static Catalog LoadDefault()
    {
        IDeserializer deserializer = new DeserializerBuilder()
            .WithNamingConvention(UnderscoredNamingConvention.Instance)
            .IgnoreUnmatchedProperties()
            .Build();

        var entities = new Dictionary<string, EntityType>(StringComparer.Ordinal);
        var entitiesFile = deserializer.Deserialize<EntityTypesFile>(Resources.EntityTypesYaml());
        foreach (EntityEntry e in entitiesFile.Entities ?? new())
        {
            var entity = new EntityType(e.Name!, e.PhileasField!, e.PhileasStrategiesField!);
            entities[entity.Name.ToUpperInvariant()] = entity;
        }

        var strategies = new Dictionary<string, Strategy>(StringComparer.Ordinal);
        var strategiesFile = deserializer.Deserialize<StrategiesFile>(Resources.StrategiesYaml());
        foreach (StrategyEntry s in strategiesFile.Strategies ?? new())
        {
            var args = new List<StrategyArg>();
            foreach (ArgEntry a in s.Args ?? new())
            {
                args.Add(new StrategyArg(a.Name!, a.PhileasField, a.Type,
                    a.EnumValues ?? new List<string>()));
            }
            var strategy = new Strategy(s.Name!, s.PhileasEnum!, args,
                s.PhileasStrategyDef == "dateFilterStrategy");
            strategies[strategy.Name.ToUpperInvariant()] = strategy;
        }

        return new Catalog(entities, strategies);
    }

    /// <summary>Returns the entity type with the given (case-insensitive) name, or null.</summary>
    public EntityType? GetEntity(string? name)
    {
        if (name is null) return null;
        return _entitiesByName.TryGetValue(name.ToUpperInvariant(), out EntityType? e) ? e : null;
    }

    /// <summary>Returns the strategy with the given (case-insensitive) name, or null.</summary>
    public Strategy? GetStrategy(string? name)
    {
        if (name is null) return null;
        return _strategiesByName.TryGetValue(name.ToUpperInvariant(), out Strategy? s) ? s : null;
    }

    /// <summary>The Phileas strategy enums classified date-only (dateFilterStrategy).</summary>
    public IReadOnlySet<string> DateOnlyStrategyEnums() =>
        _strategiesByName.Values.Where(s => s.DateOnly).Select(s => s.PhileasEnum).ToHashSet();

    /// <summary>The PhiSQL entity name for a Phileas identifier field, or null.</summary>
    public string? EntityNameForField(string field) =>
        _entitiesByName.Values.FirstOrDefault(e => e.PhileasField == field)?.Name;

    // --- YAML DTOs (snake_case via UnderscoredNamingConvention) ---------------

    private sealed class EntityTypesFile { public List<EntityEntry>? Entities { get; set; } }

    private sealed class EntityEntry
    {
        public string? Name { get; set; }
        public string? PhileasField { get; set; }
        public string? PhileasStrategiesField { get; set; }
    }

    private sealed class StrategiesFile { public List<StrategyEntry>? Strategies { get; set; } }

    private sealed class StrategyEntry
    {
        public string? Name { get; set; }
        public string? PhileasEnum { get; set; }
        public string? PhileasStrategyDef { get; set; }
        public List<ArgEntry>? Args { get; set; }
    }

    private sealed class ArgEntry
    {
        public string? Name { get; set; }
        public string? PhileasField { get; set; }
        public string? Type { get; set; }
        public List<string>? EnumValues { get; set; }
    }
}
